Applied temporary ratings and audit objects

getorymckeag opened the pull request:
The media type names I ended up with are a little long on the audit objects.  You have `slim` and `detailed` in your requirement sets.  I'm on the fence- I thought that maybe purpose should be more explicit, but it is a bit obnoxiously long.  

Oh yeah, and temporary ratings.

getorymckeag committed:
Applied temporary ratings and audit objects

catkins-miso commented on openapi.yaml:
```diff
@@ -118,6 +160,12 @@ paths:
+        This operation uses media types to control verbosity of the data fetched.  
+        The default media type (application/vnd.trolie.forecast-limit-set.v1+json) 
+        simply includes the rating data.  For applications that need it, the media type
+        application/vnd.trolie.forecast-limit-with-audit-trail-set.v1+json may be requested, 
```
You still have to explain what it is, so why try to cram a somewhat inaccurate description into a name?

After all, it's image/jxr not image/jpeg-extended-range-for-continuous-tone-photographs.

getorymckeag replied:
Yeah agreed, changed.

catkins-miso commented on openapi.yaml:
```diff
@@ -109,6 +109,48 @@ tags:
+    
+      The purpose and behavior of a temporary rating is two-fold.
+   
+      First, according to NERC, all exceptions to normal rating
```
according to FERC

getorymckeag replied:
Oops, fixed.

catkins-miso commented on openapi.yaml:
```diff
@@ -1023,6 +1141,372 @@ paths:
+              $ref: '#/components/headers/X-Rate-Limit-Remaining'
+            X-Rate-Limit-Reset:
+              $ref: '#/components/headers/X-Rate-Limit-Reset'
+            X-TROLIE-Limit-Units:
```
General feedback on this header:

The media types (input and output / proposals and snapshots) themselves should specify the units, right? So why is this header needed? If a particular TROLIE instance doesn't support the units provided in a submitted proposal because of a business decision, that's not a runtime concern per se. Yes, that TROLIE instance should rejected the proposal, but the sender would have known up front because they know the modeling situation.

Example, MISO only excepts forecasts in MVA for segments, yet only accept MW values for interfaces.

I just don't understand why this header is needed, and it is very unconventional.

getorymckeag replied:
Yes, I agree it's unconventional- but this is also an important interop usecase.  MISO, and in fact all ISOs I've been talking to, _prefer_ to only accept MVA for segments.  I say prefer, because maybe you have been able to hold your ground on this.  However, other operators I've talked to have not been able to hold their ground.  They have corner cases where, _well, on this one line, the guy that calculates that can ONLY give us amps, or MW + PF, and so we have to handle that._  I see this as being one of the compromises necessary for interop to be a reality.  The intent of the API is that if nothing is provided, then they are expected to both use and send MVA.  The alternatives (amps, MW + PF) are documented as to when they can be used.  The key idea is that TROLIE always normalizes to MVA, but the weird corner-case consumers may use the two alternative limit types if they are special.  

Could I do this with media types instead?  Technically, sure.  However, to do so, we'd have to dramatically increase the number of media types in the API- then, when we consider naming conventions, you'd have to have a separate media type for every possible permutation.  We would have an explosion in media types, which I don't think is what any of us wants.  If you've got an idea for an alternate way to solve the problem, I'm for it.  However, in order to support interop, we have to have this.  In fact, if we _don't_ allow for MW + PF in TROLIE, we are explicitly saying (at least from my current understanding) that TROLIE is explicitly incompatible with NYISO.  I don't think we want to be there.

catkins-miso replied:
I think I do have an alternative you'll find acceptable, so please indulge some more context and discussion. I finally get that you'd prefer
```http
GET /forecast
Accept: application/vnd.trolie.forecast-slim+json
X-TROLIE-Limit-Units: MVA
```
to
```http
GET /forecast
Accept: application/vnd.trolie.forecast-slim+json,mva
```

I definitely would not suggest anything that would proliferate media types. My understanding is that the existing schema supports different quantifications (scalars + units) of facility ratings.
![image](https://github.com/ge-grid-software/TROLIE/assets/57670484/dc84bfe3-f588-4419-9609-7e1002d244c2)

MISO may ultimately reject proposals for facility segments that are quantified in units other than MVA, but that is *not* a reason the TROLIE proposal representation can't afford the Ratings Provider the ability to send a proposal quantified in other units. If anything I would be open to *adding* additional options for units in the proposal schema, e.g., MVAR, if someone had the need. In fact, we need a real power only unit, i.e., MW, for interface limits in TROLIE.

I'm not sure how the header helps in submission.  I can see the peering bootstrap concept we've talked about pulling in the peer-TROLIE's profile and learning that it needs to convert to MVA before sending. Is that how you envision this working? Like the LEP client would pre-flight with a `HEAD` request, get a `X-TROLIE-Limit-Units: MVA`, then convert its MW to MVA using the present power factor value?

I'm not sure this is realistic in all cases. MVA to amps requires a known voltage, for example, but that is usually established independently via the voltage solution. Conversion between units also requires agreement on significant digits and/or truncation versus rounding.

Moreover, TROLIE submissions are considered telemetry by MISO. We expect and assume that a telemetered value is the value that the telemetering party is operating against. By extension, they need to be exacting in the value they send: If they operate in amps but MISO operates in MVA, they might use their voltage solution to determine the latter before sending. We could go on, and MISO is aware of the varying opinions on the right units to utilize.

The point is that the representation needs the flexibility to represent any appropriate units for transmission interoperability, but there are engineering decisions that precede the establishment of an exchange, precluding the possibility of a runtime discovery and conversion regime, in general.

In sum, I think the Ratings Obligation profile should specify the acceptable units for proposals per ratings provider as well as the acceptable units for the snapshot representation per monitoring area.

getorymckeag replied:
I like the qualifier idea on the media types.

catkins-miso replied:
I know I was long-winded, so I'll ask for specific feedback on my actual proposal, included below for convenience. I only mentioned the Accept header option to illustrate that I understood the objective. I do not think runtime content negotiation is sufficient to address the objective (which is why I was long-winded).

>I think the Ratings Obligation profile should specify the acceptable units for proposals per ratings provider as well as the acceptable units for the snapshot representation per monitoring area.

What do you think?

getorymckeag replied:
Oh.  
> I think the Ratings Obligation profile should specify the acceptable units for proposals per ratings provider as well as the acceptable units for the snapshot representation per monitoring area.

Yep, I probably missed that, and fully agree it's simpler than what I've done here in the API.

catkins-miso commented on openapi.yaml:
```diff
@@ -1023,6 +1141,372 @@ paths:
+              $ref: '#/components/headers/X-Rate-Limit-Reset'
+            X-TROLIE-Limit-Units:
+              $ref: '#/components/headers/X-TROLIE-Limit-Units'
+            X-TROLIE-Naming-Scheme:
```
Basically the same comment as above. I'm realizing that this PR isn't the place to hash it out, but this is important.

The identity of the segments known to that ratings provider can be specified using the CIM-style names as we discussed previously:
>However, MISO would require the TO to send `"facility-names":[{"name":"RICKROLL_LN_1","authority":"MISO","mrid":"..."}]`  Other names could be included by them. Again, I think this can be addressed by including this naming affordance in the media types instead of the `X-` header on the external submission side.

In other words, both parties would need to know the primary name used by the other ahead of time.

I don't see how this can be a runtime behavior, so a header just doesn't make sense to me.

getorymckeag replied:
Ah, this is a flavor of the interop problem with limit units, but it is even worse.  How many names does a thing have?  How many systems of names do you know about?  Do your neighbors use different names than you for the same stuff?  

I am looking at CIM style names for this, and if we can use media type qualifiers instead of headers, that would be cool.

getorymckeag replied:
Removed from API until we can readdress with wider group.

catkins-miso commented on openapi.yaml:
```diff
@@ -3274,7 +3843,67 @@ components:
+      properties:
+        _links:
+          allOf:
+          - $ref: '#/components/schemas/hal-links-block'
```
Can we figure out a general strategy to extricate the interop spec from the HAL stuff you need for the LEP UI? Or I need to get over it and prepare to sell HAL to everyone. The interop use cases do not overlap much with the UI use cases, and every vendor is going to have their own UI approach. I have no problem with using relative URL paths as unique identifiers, but I think most of the unique identifiers that are need in interop are specific to the entities doing the exchange.

getorymckeag replied:
Yeah in all cases (I think) where people have to submit IDs for things, I have included the actual IDs in addition to links, because submitting links is obviously awkward.  I'm kind of loathe to remove it now, as I know our customer's TOs are starting ask whether they consider these methods as relatively stable or not.  I haven't heard any pushback on it from other customers so far.  However, to be fair, I also don't think there is a deep understanding of the architectural pattern.  It seems like most users just want to understand what they have to send, and other than that have no comments.  I feel like HAL is useful long-term, although I understand you driving towards minimizing the scope.  

I don't know, just ruminating some thoughts out loud here.  Willing to listen.

catkins-miso replied:
I fully support HAL in LEP; its use cases will absolutely benefit from a uniform approach to HATEOAS representations. But those aren't TROLIE's interop use cases for the most part. The TROLIE media types are basically as subset, the resource state part. It's worth changing the media type names to reflect that. Your LEP customers can enjoy the uniformity of the HAL envelope and tolerate minor changes to the resource representation schema as both LEP and TROLIE evolve.

getorymckeag replied:
Removed HAL from TROLIE.

catkins-miso commented on openapi.yaml:
```diff
@@ -3274,7 +3843,67 @@ components:
+          $ref: '#/components/schemas/timestamp'
+        values:
+          $ref: '#/components/schemas/limit-value-set'
+        temporary-aar-exception:
```
Why not implement these as separate endpoints? We need to be able to validate them differently. We need interface limits.

Also, this allows for a representation where temporary-aar-exception = true and temporary-seasonal-rating = true.

getorymckeag replied:
Maybe this a semantic difference in understanding on my part, but I don't actually see them as being different.  I also see this:
temporary-aar-exception = true and temporary-seasonal-rating = true

as being valid.

catkins-miso replied:
They are implemented very differently in the clearinghouse. temp AAR exception *must* be proposed by LEP if it doesn't have a telemetered forecast, since it should supersede temp table use. 
![image](https://github.com/ge-grid-software/TROLIE/assets/57670484/2641ad7c-9268-49c8-8ff1-1385f36274d6)

Also, there is some uncertainty around the proper validation for AAR exceptions versus temporary seasonal ratings. In particular, we might want to allow the temporary seasonal to go lower than the reasonability bounds defined on the equipment.

Finally, only equipment with a proper ratings obligation can have a temp AAR exception. Any modeled equipment can have seasonal.

catkins-miso reviewed (commented):
some initial thoughts

getorymckeag commented:
> some initial thoughts

You know, something funny here, I meant to actually submit this PR against the new repo- I pushed a branch and everything.  Just user error.  Since we started a conversation here, I'll just finish it here.

getorymckeag commented:
@catkins-miso I added a couple more commits that I'm hoping get us closer to the cleaner starting point.  
- Naming scheme, units headers gone
- Geographic regions gone
- HAL gone (and paging, by extension)
- Split temporary ratings into AAR exceptions and temporary seasonal ratings.  Note that the schema is the same- saw no reason to copy-paste that, although I did create different media types for now.  

If you're willing to merge this in, I think this will give me a better starting point for documenting examples for TOs.  There are are still a few things I'd like to clean up later, but I think they can wait until later discussions.  Just as food for thought:
- The model I have for seasons and seasonal ratings has... got a number of things wrong with it, and they should probably be read-only in TROLIE.  I'll get there.  
- Without the use of HAL, I think the queries for segments are probably noise (nothing needs to "link" to a segment really).  Although I do believe the GETs for the transmission facilities visible to TROLIE are likely still needed for interop.  
- I kind of think all the stuff around "missing ratings" can probably go.  Is it interesting?  Of course.  I think the interop value is thin though.

catkins-miso commented:
Closing this, the referenced changes are in #10
